proposals/0014-wave-size-range.md

```diff
@@ -0,0 +1,283 @@
+Consequently, if more than one target wave size is of interest, separate entry
+ points will have to be created for each wave size.
+These will each have to be compiled and shipped separately and, at runtime,
+ the appropriate shaders will have to be selected.
+This increases the size of the shipped product and potentially slows down
+ runtime shader loading.
+
+Ranges of wave sizes are already present in areas of the D3D API.
+The query to determine what wave sizes are supported takes a range and similar
+ information is transmitted from the shader to the runtime, though not the
+ driver.
+
+## Proposed solution
+
+Modifying the parameters of the `WaveSize` attribute that was introduced by
+ Shader Model 6.6 to take additional parameters will allow the shader author to
+ specify fully what the shader supports and what it prefers.
+
+By adding an optional second `WaveSize` parameter, shader authors can provide
+ two values that represent the range of wave sizes that the shader can support.
+This allows the same shader to be used in the event of the availability of a
+ wave size that the shader has optimized for specifically as well as other sizes
+ that are supported, but might not be so optimal.
+
+Some platforms that support a range of wave sizes might overlap with the range
+ specified as supported by the shader in more than one value.
+In these cases, the runtime has a choice of wave sizes that it probably doesn't
+ have the information needed to choose the best wave size.
+To give the runtime the needed information, it is useful to specify the
+ preferred wave size in addition to the full range of acceptable values.
+By adding an optional third `WaveSize` parameter, shader authors can specify the
+ optimal wave size for a given shader in addition to the range of acceptable
+ sizes.
+
+## Detailed Design
+
+### HLSL additions
+
+The existing `WaveSize` attribute gains a second overload:
+
+```HLSL
+[WaveSize(<minWaveSize>, <maxWaveSize>, [prefWaveSize])]
+void main() ...
+```
+
+Where `minWaveSize` is the minimum wave size supported by the shader
+ representing the beginning of the allowed range,
+ `maxWaveSize` is the maximum wave size supported by the shader
+ representing the end of the allowed range,
+ and `prefWaveSize` is the optional preferred wave size representing the size
+ expected to be the most optimal for this shader.
+
+### DXIL Additions
+
+The existing `WaveSize` value is stored as a metadata scalar constant.
+To store the additional values, an additional metadata tag indicating a tuple of
+ three values representing the parameters to `WaveSize` given by the user.
+In the case where no preferred wave size is specified, the value zero will
+ indicate that nothing was specified.
+
+### SPIR-V Additions
+
+To make use of this attribute, SPIRV would have to represent it in new SPIRV
+ operations that drivers can read to determine what the values are an how to
+ use to them to choose the best wave size for a given shader.
```

Comment:
To be more explicit here. A new `ExecutionMode` value would need to be defined for the `OpExecutionMode` instruction to allow specifying `SubgroupSize` with 3 operands instead of 1.

I did a quick scan and I don't see a SPIR-V extension that supports this today, but it should be pretty straightforward. It is probably worth clearly stating that the SPIR-V extension is not planned as part of the work for implementing this feature.

Reply:
Vulkan would handle the old WaveSize attribute on the API side and not in SPIR-V at all. `WaveSize(x)` would be best translated creating a pipeline with a required subgroup size and (likely) the full subgroups flag. So it is little incongruent to move part of it into SPIR-V. It could certainly be done, but there would be some awkwardness.

Reply:
@alan-baker, I think you're misunderstanding. In Vulkan the API allows you to specify the save size while creating the device, that's somewhat orthogonal to the HLSL `WaveSize` attribute. The existing `WaveSize` attribute should translate to the `OpExecutionMode` `SubgroupSize` property which is part of Vulkan 1.1 SubgroupDispatch.

This allows the shader to specify on an entry point the required subgroup size. That prevents the shader from being used if the Subgroup isn't the required size, which allows the author to make safe assumptions about subgroup size.

The extension here is to allow a shader to specify a range of subgroup sizes. That allows a shader author to make assumptions about the subgroup size within a specified range, which can allow them to ship less shader variants when needing to support specific subgroup sizes for optimal performance.

In both cases the requirement needs to be captured in the byte code to enforce that the shader cannot be used with a mismatched subgroup size.

Reply:
Vulkan's functionality isn't about device creation. The subgroup size control is part of pipeline creation. There is a device property that specifies the min and max sizes supported by the device, but at pipeline creation time you can require a specific power-of-two size within that range (using [VkShaderStageRequiredSubgroupSizeCreateInfo](https://registry.khronos.org/vulkan/specs/1.3-extensions/html/vkspec.html#VkPipelineShaderStageRequiredSubgroupSizeCreateInfo)). 

Additionally, there are pipeline creation flags for compute-like shaders provided:
* `VK_SHADER_CREATE_ALLOW_VARYING_SUBGROUP_SIZE_BIT_EXT`: subgroup size can vary (incompatible with a required size)
* `VK_SHADER_CREATE_REQUIRE_FULL_SUBGROUPS_BIT_EXT`: require the shader to be launched with no inactive invocations.

This functionality was made core as part of Vulkan 1.3 (see [subgroup size control](https://registry.khronos.org/vulkan/specs/1.3-extensions/html/vkspec.html#VkPhysicalDeviceSubgroupSizeControlProperties)).

The SubgroupDispatch functionality came to SPIR-V from OpenCL and is not currently supported by Vulkan implementations. You can see this be the eventual dependence on the `Kernel` capability (via `DeviceEnqueue`).

Again, there could be a Vulkan/SPIR-V extension to allow some of that functionality for Vulkan, but it doesn't fit the current way Vulkan handles it. I understand wanting to capture the HLSL attribute as part of the compiled shader.

Reply:
I don't think that changes what we need to do here for HLSL. If Vulkan doesn't support a shader requiring a specific subgroup size, then Vulkan extensions will be required both for supporting the existing SM 6.6 feature, and the SM 6.8 feature.

For SM 6.6's feature the SPIR-V SubgroupSize execution mode captures the required information (whether or not Vulkan supports the required capability, it is at least defined in SPIR-V). For SM 6.8 compatibility an additional SPIR-V execution mode (and capability, extension etc) will be required.

We're not actively driving changes in Vulkan or SPIR-V. We've included information in this proposal as a guide to anyone who might want to drive those efforts.